src/parser/core/modules/BuffWindow.tsx

```diff
@@ -0,0 +1,329 @@
+				return Math.ceil((windowDurationMillis - fightTimeRemaining) / gcdEstimate)
+			}
+		}
+
+		// Default: no rushing reduction
+		return 0
+	}
+
+	// For consumers that have tracked cooldowns that expect the same number of that cooldown per window, this will use the
+	//   expectedPerWindow property on that cooldown as the baseline.  This method can be overridden if the logic of
+	//   expected tracked cooldowns per window is variable
+	protected getBaselineExpectedTrackedCooldown(buffWindow: BuffWindowState, cooldown: BuffWindowTrackedCooldown): number {
+		return cooldown.expectedPerWindow || 0
+	}
+
+	// Override point for class-specific logic to change expected uses of a tracked cooldown per BuffWindow - default no effect
+	// Return a positive number to increase expected GCDs for this window, and a negative number to decrease
+	protected changeExpectedTrackedCooldownClassLogic(buffWindow: BuffWindowState, cooldown: BuffWindowTrackedCooldown): number {
+		return 0
+	}
+
+	private getBuffWindowExpectedGCDs(buffWindow: BuffWindowState): number {
+		return this.getBaselineExpectedGCDs(buffWindow) + this.changeExpectedGCDsClassLogic(buffWindow) - this.reduceExpectedGCDsEndOfFight(buffWindow)
+	}
+
+	private getBuffWindowExpectedTrackedCooldowns(buffWindow: BuffWindowState, cooldown: BuffWindowTrackedCooldown): number {
+		return this.getBaselineExpectedTrackedCooldown(buffWindow, cooldown) + this.changeExpectedTrackedCooldownClassLogic(buffWindow, cooldown)
+	}
+
+	private onComplete() {
+		if ( this.buffAction && this.expectedGCDs ) {
+			const missedGCDs = this.buffWindows
+				.reduce((sum, curWindow) => {
+					const expectedGCDs = this.getBuffWindowExpectedGCDs(curWindow)
+					return sum + Math.max(0, expectedGCDs - curWindow.gcds)
+				}, 0)
+
+			this.suggestions.add(new TieredSuggestion({
+				icon: this.buffAction.icon,
+				content: this.expectedGCDs.suggestionContent,
+				tiers: this.expectedGCDs.severityTiers,
+				value: missedGCDs,
+				why: <Trans id="core.buffwindow.suggestions.missedgcd.why">
+					{missedGCDs} <Plural value={missedGCDs} one="GCD was" other="GCDs were" /> missed during {this.buffAction.name} windows.
+				</Trans>,
+			}))
+		}
+
+		if ( this.buffAction && this.requiredGCDs ) {
+			const allowedGCDsById = this.requiredGCDs.actions.map(a => a.id)
+			const invalidGCDs = this.buffWindows
+				.reduce((sum, curWindow) => sum + Math.max(0, curWindow.gcds - curWindow.getActionCountByIds(allowedGCDsById)), 0)
+
+			this.suggestions.add(new TieredSuggestion({
+				icon: this.requiredGCDs.iconAction.icon,
+				content: this.requiredGCDs.suggestionContent,
+				tiers: this.requiredGCDs.severityTiers,
+				value: invalidGCDs,
+				why: <Trans id="core.buffwindow.suggestions.badgcd.why">
+					{invalidGCDs} incorrect <Plural value={invalidGCDs} one="GCD was" other="GCDs were" /> used during {this.buffAction.name} windows.
+				</Trans>,
+			}))
+		}
+
+		if ( this.trackedCooldowns ) {
```

Comment:
If you initialise these as `protected foo: Foo[] = []` rather than `protected foo?: Foo[]`, it'll remain an optional property, but you'll be able to skip the existence checks as it always has _some_ value.